### PR TITLE
porting and adding Falabracman.activity

### LINKED ISSUE #206: Port the Falabracman Sugar activity
Port the [Falabracman](http://activities.sugarlabs.org/en/sugar/addon/4215) Sugar activity (Python) to Sugarizer (JavaScript).
Follow the Sugarizer tutorial "[Create your own activity](https://github.com/llaske/sugarizer/blob/dev/docs/tutorial.md)" to create a new Falabracman activity.

Guidelines:

-  The activity should be correctly integrated into Sugarizer (the activity icon should adapt its color when launched from the Sugarizer home view)
-  The activity should be responsive (work on any screen size)
-  The activity should work with the keyboard and with the mouse (to support touch screen)
- The activity should use localization library lib/webL10n.js
- The activity should work on any browser (Chrome, Firefox, Safari and EDGE)
-  The activity should not use Internet connection (link to local content only)
- The activity should implement all existing features of the original activity
- The activity should retrieve graphics elements (icons, images, sounds, ...) of the original activity

### PR DESCRIPTION
**Pull request by prakashujjwal1010**
- fixes #206 
- activity interated into sugarizer successfully
- activity works with keyboard, mouse (also supports touch event)
- activity is localized
- implements local all existing features and retrieve all graphics elements of activity(python version) 
- activity is responsive
- used datastore to store game context into journal
- added sound
- added a restart toolbar button (which is not present in original one)
- added attribution for icon used or restart button

ezgif com-video-to-gif (3)

 Screenshot from 2020-01-28 21-59-02

Screenshot from 2020-01-28 22-09-29

Screenshot from 2020-01-28 22-31-27

### CODE REVIEW

**Comment by llaske** [comment]
Good job!
That's a very good start.
It's very cool to have this game in Sugarizer.

Here are my remarks:
* [ ] The window is responsive but images proportion are not preserved
* [ ] Playing with mouse/touch don't work
* [ ] Will be better to include the word list in locale.ini so it could be translated in any language
* [ ] The language of the game should be the current Sugarizer language
* [ ] Due to previous point, the menu options could be removed
* [ ] During the game the menumusic22.ogg should be played in background
* [ ] When leaving from the home and reopening the game the screen is blank
* [ ] Replace the back icon by a home icon
* [ ] A tutorial should be added
* [ ] A fullscreen button should be added

**Comment by prakashujjwal1010** [comment]
- [x] The window is responsive as well as images proportion are also preserved 
- [x] Playing with mouse/touch working ( it was already working ,user have to drag his mouse just like in subway surfers game)
- [x]  the menu options could be removed
- [x] word list added in locale.ini
- [x] During the game the menumusic22.ogg should be played in background
- [x] When leaving from the home and reopening the game the screen is blank ( fixed )
- [x] When the game is lost the user should go back to the home
- [x] Replace the back icon by a home icon( attribution also added for new icon)
- [x]  A tutorial should be added
- [x]  A fullscreen button should be added

**Comment by llaske** [comment]
Cool. It's better.

Few remarks:

- Regarding responsive, I think image are still flattened. See following screen capture. To the right the original image, to the left the image during the game:

image

- When a word is completed a new word is displayed but the number of life is reinitialized to the maximum. It should not.

- The value of name in activity.info should be Falabracman else the activity palette is incorrect.

image

- When the window is resized after a click on fullscreen/unfullscreen, a scrollbar appears on the whole window.

image

**Comment by prakashujjwal1010** [comment]
- [x] resizing issues fixed
- [x] no of lives persisted after a word is completed
- [x] name field corrected in activity.info

Screenshot from 2020-02-03 02-31-43

File: activities/Falabracman.activity/activity/activity.info [hunk 1]
```diff
@@ -1,5 +1,5 @@
 [Activity]
-name =  Activity
+name =  Falabracman Activity
```
Comment (llaske) [inline activities/Falabracman.activity/activity/activity.info]: Should be "Falabracman" only else the menu is incorrect:

image

**Comment by llaske** [comment]
Very nice.

BTW I think there is an issue: the speed on small screen is different from the speed on large screen. So the game is more difficult on small screen. I guess it comes from the fact than the distance at each step is not linked to the size of the screen.

Plus, I think that teachers will ask a feature to add their own dictionary to allow students to play with words they learn in classroom.
So I suggest to add again a "Settings" screen where user could add/update/remove easily words in the activity.

**Comment by prakashujjwal1010** [comment]
- the speed has been linked with canvas size
- new setting screen is added
- tutorial added for new setting screen
Screenshot from 2020-02-09 01-43-34

the custom dictionary created by  the user is stored in datastore also.
the user can add / edit / delete words in his custom dictionary

**Comment by llaske** [comment]
Nice. It does the job.
BTW the UI is slightly complex and due to button size (very small), it will be very difficult to use on a touch screen. May be you could remove the Falabracman image to get some space? Will be better too to use icons instead of text for actions (see Exerciser activity for inspiration).

It seems that when the Sugarizer language is changed, the custom dictionary is lost. I suggest than when the user choose to customize the dictionary, it become the unique dictionary whatever the language is.

Caution: with this new settings option, the menu no longer fit with small screens.

image

Plus, could you remove log messages in the console?

**Comment by prakashujjwal1010** [comment]
okay, will do .
also, by actions you mean all the buttons in setting screen or just buttons in dictionary section ?

**Comment by llaske** [comment]
Cool. I mean all buttons in this screen.

**Comment by prakashujjwal1010** [comment]
- icons added and ui improved
- menu fit issues fixed

Screenshot from 2020-02-11 02-50-24

**Comment by llaske** [comment]
Very good.

BTW I've tested on mobile devices and I've some issues:
- You're using ES6 syntax ("let", "[...", ...) in your source code and it's not recognize on some devices (specifically iOS). Could you fix it?
- All sounds are in OGG format, only MP3 format is supported on some devices (again, iOS). Could you convert it in MP3 files?

**Comment by prakashujjwal1010** [comment]
okay, i'll do it.

**Comment by prakashujjwal1010** [comment]
@llaske please review.

**Comment by llaske** [comment]
@prakashujjwal1010 I'm testing it on different platforms but it seems to work perfectly now :+1:

**Comment by llaske** [comment]
I've detected a new issue.

Steps to reproduce:

* Set Sugarizer language to Spanish
* Launch a new Falabracman instance
* Click Jugar
* Go to home
* Quit activity
* Relaunch the activity
* Click Jugar

Then the game don't start, the board is continuously redrawn.

@prakashujjwal1010  Could you fix it?

**Comment by prakashujjwal1010** [comment]
yes, i got it. it was bug in word generation for non-english languages.
it is now fixed.

**Comment by llaske** [comment]
Thanks it fixed the issue.

BTW there is another one regarding localization.
Sometimes the home screen is displayed before the locale.ini file is load so the home screen is displayed in English instead of the locale. Worse, if the network is slow (for example choose Fast 3G in Chrome Network sheet), there is an error in the console:

```
game.js:207 Uncaught TypeError: Failed to execute 'drawImage' on 'CanvasRenderingContext2D': The provided value is not of type '(CSSImageValue or HTMLImageElement or SVGImageElement or HTMLVideoElement or HTMLCanvasElement or ImageBitmap or OffscreenCanvas)'
    at Game.drawHomeScreen (game.js:207)
    at Game.main (game.js:78)
    at activity.js:105
    at sugarizer.js:184
    at IDBRequest.request.onsuccess (sugarizer.js:412)
Game.drawHomeScreen @ game.js:207
Game.main @ game.js:78
(anonymous) @ activity.js:105
(anonymous) @ sugarizer.js:184
request.onsuccess @ sugarizer.js:412
IndexedDB (async)
indexedDB_getValue @ sugarizer.js:404
(anonymous) @ sugarizer.js:422
request.onsuccess @ sugarizer.js:360
IndexedDB (async)
html5indexedDB.load @ sugarizer.js:351
html5indexedDB.getValue @ sugarizer.js:418
DatastoreObject.loadAsText @ sugarizer.js:178
(anonymous) @ activity.js:75
(anonymous) @ env.js:34
setTimeout (async)
env.getEnvironment @ env.js:33
(anonymous) @ activity.js:53
execCb @ require.js:1604
check @ require.js:844
(anonymous) @ require.js:1084
(anonymous) @ require.js:129
(anonymous) @ require.js:1127
each @ require.js:57
emit @ require.js:1126
check @ require.js:898
enable @ require.js:1114
init @ require.js:757
(anonymous) @ require.js:939
(anonymous) @ require.js:129
(anonymous) @ require.js:1127
each @ require.js:57
emit @ require.js:1126
check @ require.js:898
enable @ require.js:1114
init @ require.js:757
(anonymous) @ require.js:963
(anonymous) @ require.js:129
domReady @ domReady.js:106
domReady.load @ domReady.js:122
(anonymous) @ require.js:1044
(anonymous) @ require.js:129
on @ require.js:494
callPlugin @ require.js:915
fetch @ require.js:793
check @ require.js:823
enable @ require.js:1114
(anonymous) @ require.js:956
(anonymous) @ require.js:129
(anonymous) @ require.js:1127
each @ require.js:57
emit @ require.js:1126
check @ require.js:898
enable @ require.js:1114
init @ require.js:757
callGetModule @ require.js:1141
completeLoad @ require.js:1498
onScriptLoad @ require.js:1625
load (async)
req.load @ require.js:1830
load @ require.js:1593
load @ require.js:803
fetch @ require.js:793
check @ require.js:823
enable @ require.js:1114
enable @ require.js:1465
callPlugin @ require.js:1047
fetch @ require.js:793
check @ require.js:823
enable @ require.js:1114
enable @ require.js:1465
(anonymous) @ require.js:1099
(anonymous) @ require.js:129
each @ require.js:57
enable @ require.js:1061
init @ require.js:757
(anonymous) @ require.js:1380
setTimeout (async)
req.nextTick @ require.js:1709
localRequire @ require.js:1369
requirejs @ require.js:1691
(anonymous) @ activity.js:4
execCb @ require.js:1604
check @ require.js:844
(anonymous) @ require.js:1084
(anonymous) @ require.js:129
(anonymous) @ require.js:1127
each @ require.js:57
emit @ require.js:1126
check @ require.js:898
(anonymous) @ require.js:1084
(anonymous) @ require.js:129
(anonymous) @ require.js:1127
each @ require.js:57
emit @ require.js:1126
check @ require.js:898
(anonymous) @ require.js:1084
(anonymous) @ require.js:129
(anonymous) @ require.js:1127
each @ require.js:57
emit @ require.js:1126
check @ require.js:898
enable @ require.js:1114
init @ require.js:757
(anonymous) @ require.js:939
(anonymous) @ require.js:129
(anonymous) @ require.js:1127
each @ require.js:57
emit @ require.js:1126
check @ require.js:898
enable @ require.js:1114
init @ require.js:757
(anonymous) @ require.js:963
(anonymous) @ require.js:129
finishLoad @ text.js:152
(anonymous) @ text.js:188
xhr.onreadystatechange @ text.js:275
XMLHttpRequest.send (async)
text.get @ text.js:283
load @ text.js:187
(anonymous) @ require.js:1044
(anonymous) @ require.js:129
on @ require.js:494
callPlugin @ require.js:915
fetch @ require.js:793
check @ require.js:823
enable @ require.js:1114
(anonymous) @ require.js:956
(anonymous) @ require.js:129
(anonymous) @ require.js:1127
each @ require.js:57
emit @ require.js:1126
check @ require.js:898
enable @ require.js:1114
init @ require.js:757
callGetModule @ require.js:1141
completeLoad @ require.js:1498
onScriptLoad @ require.js:1625
load (async)
req.load @ require.js:1830
load @ require.js:1593
load @ require.js:803
fetch @ require.js:793
check @ require.js:823
enable @ require.js:1114
enable @ require.js:1465
callPlugin @ require.js:1047
fetch @ require.js:793
check @ require.js:823
enable @ require.js:1114
enable @ require.js:1465
(anonymous) @ require.js:1099
(anonymous) @ require.js:129
each @ require.js:57
enable @ require.js:1061
init @ require.js:757
callGetModule @ require.js:1141
completeLoad @ require.js:1498
onScriptLoad @ require.js:1625
load (async)
req.load @ require.js:1830
load @ require.js:1593
load @ require.js:803
fetch @ require.js:793
check @ require.js:823
enable @ require.js:1114
enable @ require.js:1465
(anonymous) @ require.js:1099
(anonymous) @ require.js:129
each @ require.js:57
enable @ require.js:1061
init @ require.js:757
callGetModule @ require.js:1141
completeLoad @ require.js:1498
onScriptLoad @ require.js:1625
load (async)
req.load @ require.js:1830
load @ require.js:1593
load @ require.js:803
fetch @ require.js:793
check @ require.js:823
enable @ require.js:1114
enable @ require.js:1465
(anonymous) @ require.js:1099
(anonymous) @ require.js:129
each @ require.js:57
enable @ require.js:1061
init @ require.js:757
callGetModule @ require.js:1141
completeLoad @ require.js:1498
onScriptLoad @ require.js:1625
load (async)
req.load @ require.js:1830
load @ require.js:1593
load @ require.js:803
fetch @ require.js:793
check @ require.js:823
enable @ require.js:1114
enable @ require.js:1465
(anonymous) @ require.js:1099
(anonymous) @ require.js:129
each @ require.js:57
enable @ require.js:1061
init @ require.js:757
(anonymous) @ require.js:1380
setTimeout (async)
req.nextTick @ require.js:1709
localRequire @ require.js:1369
requirejs @ require.js:1691
(anonymous) @ loader.js:8
Show 64 more frames
```

l guess it's because you're not waiting for the `localized` message like explained [here](https://github.com/llaske/sugarizer/blob/dev/docs/tutorial_step5.md#set-strings-value-depending-on-the-language).

**Comment by prakashujjwal1010** [comment]
@llaske , i am not able to reproduce the error as the activity is working fine even in slow net connection fast 3g or slow 3g or custom ( 10 kbps). however, by looking at the error , i have made some changes. could you check it now?

**Comment by llaske** [comment]
Works now. Thanks.

**Comment by llaske** [comment]
And thanks a lot for this great work!

**Pull request merged**